node-registrar/pkg/server/handlers.go

```diff
@@ -0,0 +1,336 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/gin-gonic/gin"
+	"github.com/threefoldtech/tfgrid-sdk-go/node-registrar/pkg/db"
+)
+
+// @Summary list farms
+// @Description list farms with specific filter
+// @Accept  json
+// @Produce  json
+// @Param farm_name query string false "farm name"
+// @Param farm_id query uint64 false "farm id"
+// @Param twin_id query uint64 false "twin id"
+// @Param page query int false "Page number"
+// @Param size query int false "Max result per page"
+// @Success 200 {object} []db.Farm
+// @Failure 400 {object} error
+// @Router /farms/ [get]
+func (s Server) listFarmsHandler(c *gin.Context) {
+	var filter db.FarmFilter
+	limit := db.DefaultLimit()
+
+	err := parseQueryParams(c, &limit, &filter)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	farms, err := s.DB.ListFarms(filter, limit)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{
+		"farms": farms,
+	})
+}
+
+// @Summary get farm
+// @Description get a farm with specific id
+// @Accept  json
+// @Produce  json
+// @param farm_id path uint64 true "farm id"
+// @Success 200 {object} db.Farm
+// @Failure 404 {object} db.ErrRecordNotFound
+// @Failure 400 {object} error
+// @Router /farm/{farm_id} [get]
+func (s Server) getFarmHandler(c *gin.Context) {
+	farmID := c.Param("farm_id")
+
+	id, err := strconv.ParseUint(farmID, 10, 64)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": fmt.Sprintf("Invalid farm_id: %v", err.Error())})
+		return
+	}
+
+	farm, err := s.DB.GetFarm(id)
+	if err != nil {
+		if errors.Is(err, db.ErrRecordNotFound) {
+			c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
+			return
+		}
+
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{
+		"farm": farm,
+	})
+}
+
+// @Summary create a farm
+// @Description creates a farm
+// @Accept  json
+// @Produce  json
+// @Param farm_id body uint64 true "farm id"
+// @Param farm_name body uint64 true "farm name"
+// @Param twin_id body uint64 true "twin id"
+// @Param dedicated body bool false "dedicated farm"
+// @Param farm_free_ips body uint64 false "farm free ips"
+// @Success 200 {object} db.Farm
+// @Failure 400 {object} error
+// @Router /farms/ [post]
+func (s Server) createFarmHandler(c *gin.Context) {
+	var farm db.Farm
+
+	if err := c.ShouldBindJSON(&farm); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": fmt.Sprintf("failed to parse farm info: %v", err.Error())})
+		return
+	}
+
+	err := s.DB.CreateFarm(farm)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{
+		"message": "Farm created successfully",
+		"farm":    farm,
+	})
+}
```

Comment:
The primary issues with these handlers are as follows:

- Missing API Authorization and Authentication: There is no mechanism to restrict or control access, allowing anyone to perform actions on any resource. For example, one user can update another user's farm without restrictions.
- Lack of Validation: The handlers do not validate input or handle missing fields, which increases the likelihood of invalid or incomplete data being stored in the database.

Reply:
I added auth middleware to check if the twin sending the request is the same as the node owner, but might need to discuss how twins are represented/managed and update the auth accordingly 

need to add more validation too but didn't get the time to do so